File: leetcode/Python/search-suggestions-system.py
class Solution:
    def __init__(self):
        self.v = []
        self.next = None
    def insert(self, i, word):
        cur, idx = self, 0
        for ch in word:
            idx = ord(ch)-ord('a')
            if cur.next is None:
                cur.next = [Solution() for i in range(26)]
            cur.next[idx].v.append(i)
            cur=cur.next[idx]
    def suggestedProducts(self, products: List[str], searchWord: str) -> List[List[str]]:
        sol,idx,cur,sz = [], 0, self, 0
        products.sort()
        for i, v in enumerate(products):
            self.insert(i, v)
        for ch in searchWord:
            m = []
            idx = ord(ch)-ord('a')
            if cur.next is None:
                sol.append(m)
                continue
            cur=cur.next[idx]
            for i in range(min(3, len(cur.v))):
                m.append(products[cur.v[i]])
            sol.append(m)
        return sol

File: leetcode/C++/search-suggestions-system.cpp
class Solution {
public:
    vector<int> v;
    Solution *t;
    Solution(){t=NULL;}
    Solution* makeTrie(){return new Solution[26];}
    void insert(int &i, string &word){
        Solution *cur=this;
        int idx=0;
        this->v.push_back(i);
        for(char &ch : word){
            idx=ch-'a';
            if(cur->t==NULL) cur->t=makeTrie();
            cur->t[idx].v.push_back(i);
            cur=&cur->t[idx];
        }
    }
    vector<vector<string>> suggestedProducts(vector<string>& products, string searchWord) {
        vector<vector<string>> sol;
        Solution *cur = this;
        sort(products.begin(),products.end());
        for(int i=0; i<products.size(); ++i) insert(i,products[i]);
        for(char &ch : searchWord){
            vector<string> v;
            if(cur->t==NULL) {
                sol.push_back(v);
                continue;
            }
            cur=&cur->t[ch-'a'];
            for(int i=0; i<3 && i<cur->v.size(); ++i){
                v.push_back(products[cur->v[i]]);
            }
            sol.push_back(v);
        }
        return sol;
    }
};